### DISCUSSION TITLE
Can I decline the connection in the upgrade handler?

### DISCUSSION THREAD

**Discussion by Fxlr8** [uNetworking/uWebSockets.js#367] [Q&A] [answered]
In the latest update the `req` is no longer passed to the `open` handler. Previously I used `open` handler to check cookies and implement the authorization. I could call `ws.close()` to close the connection immediately. Is it possible to prevent the connection in the `upgrade` handler?

**Comment by ghost** [comment, marked as answer]
It's not req.end it is res.end.

So yes you can do res.end or res.close while checking headers in upgrade handler

**Comment by ghost** [comment]
Also it's not correct that you could immediately close the connection before. You only had ws and req so the http response had already been sent.

There are examples for this

https://github.com/uNetworking/uWebSockets.js/blob/master/examples/Upgrade.js

**Comment by Fxlr8** [comment]
Oops, you are right there were a couple of mistakes in my question. 
1. Of course I meant `res.end()`.
2. I used `ws.close()` metod in the `open` handler to close it immediately

Now about the `upgrade` handler:
The `res.close()` method is too aggressive. It causes `ECONNRESET` error that does not give any useful information to the client.
```
res.writeStatus('401 Unauthorized')
res.end()
```
is a little better, but it causes `Uncaught Error: Unexpected server response: 401` on the client. 

Is there a way for server to say "pls go away I don't want to connect with you" which is expected by the client?

**Comment by ghost** [comment]
I don't think so. That's something you have to handle on the client side.

**Comment by hst-m** [comment]
If you want to send a [code](https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent)/reason and prevent the error you have to do it in the open handler with `ws.end(code, reason)`. If you close the connection in the upgrade handler it will create the error event on the client and you have to handle the response without knowing the reason. The http 401 code only displays in the console and is [not accessible](https://stackoverflow.com/a/50685387) with code
